### PR TITLE
fix post cohort endpoint to accept name and users

### PR DESCRIPTION
**Pull request by MyrtheDullaart**
The POST cohort endpoint only accepted a start and end date. This has now been fixed to also accept a cohort name and list of users.

### CODE REVIEW

**Review by LeonardoSaraceli — APPROVED** [review]

**Review by FBagdeli — APPROVED** [review]

File: src/utils/validationFunctions.js [hunk 1]
```diff
@@ -72,3 +72,21 @@ export function validateCanModify(targetUserId, user) {
   }
   return true
 }
+
+export function validateCreateCohort(name, users, startDate, endDate) {
+  if (!name) {
```
Comment (dogezen) [inline src/utils/validationFunctions.js:77]: can you accept an empty name? ie name === ""?

File: src/domain/cohort.js [hunk 2]
```diff
@@ -4,19 +4,44 @@ import dbClient from '../utils/dbClient.js'
...
+          }
+        }
+      },
       deliveryLogs: true
     }
   })
 
-  return new Cohort(createdCohort)
+  if (createdCohort) {
+    return { ...createdCohort, users: flatten(createdCohort) }
```
Comment (dogezen) [inline src/domain/cohort.js:41]: a bit confused about this: can you explain the `users: flatten(createdCohort)` what does this exactly do to the createdCohort?

File: src/controllers/cohort.js [hunk 3]
```diff
@@ -2,26 +2,29 @@ import { createCohort, getAllCohorts, getCohortById } from '../domain/cohort.js'
...
-  const { startDate, endDate } = req.body
+  const { name, users, startDate, endDate } = req.body
 
-  if (!startDate || !endDate) {
-    return sendMessageResponse(res, 400, { error: ERR.DATE_REQUIRED })
-  }
-  if (!dateRegex.test(startDate) || !dateRegex.test(endDate)) {
-    return sendMessageResponse(res, 400, { error: ERR.DATE_FORMATTING })
+  try {
+    validation.validateCreateCohort(name, users, startDate, endDate)
```
Comment (dogezen) [inline src/controllers/cohort.js:10]: I imagine users is going to be an array of IDs -> how are you validating if the IDs actually correspond to valid existing users in your DB?

**Review by dogezen — CHANGES_REQUESTED** [review]
Please see comments

File: src/domain/cohort.js (hunk 2 above)
Comment (MyrtheDullaart) [reply src/domain/cohort.js:41]: Of course! The problem is that the user always returns a user that has the profile nested. However, the front end prefers to not have it nested so we need to take the nested profile out. I do this with the flatten function and then return the created cohort, with the users (with the nested profile) replaced with the non nested users. Hope that make sense! Let me know if you have further questions!

File: src/utils/validationFunctions.js (hunk 1 above)
Comment (MyrtheDullaart) [reply src/utils/validationFunctions.js:77]: No this should not be accepted. When I run it now with a name of "", we get an error back that name is required!

File: src/controllers/cohort.js (hunk 3 above)
Comment (MyrtheDullaart) [reply src/controllers/cohort.js:10]: This is done in this function: 
```js
export async function createCohort(name, users, startDate, endDate) {
  const createdCohort = await dbClient.cohort.create({
    data: {
      startDate,
      endDate,
      name,
      users: {
        connect: users.map((user) => {
          return { id: user.id }
        })
      }
    },
    include: {
      users: {
        select: {
          id: true,
          cohortId: true,
          role: true,
          email: true,
          profile: {
            select: {
              firstName: true,
              lastName: true,
              bio: true,
              githubUsername: true
            }
          }
        }
      },
      deliveryLogs: true
    }
  })

  if (createdCohort) {
    return { ...createdCohort, users: flatten(createdCohort) }
  }

  return createdCohort
}
```

More specifically, in this part: 
```js
users: {
        connect: users.map((user) => {
          return { id: user.id }
        })
      }
```

Here it only connects the user to the created cohort if the users with the provided id exists.

Let me know if I'm missing something here!

Comment (dogezen) [reply src/controllers/cohort.js:10]: OK, so you are simply catching any thrown errors of inexistent user id - ok that's fine by me

File: src/domain/cohort.js (hunk 2 above)
Comment (dogezen) [reply src/domain/cohort.js:41]: OK - got it, flatten is a custom function that does that for you.

**Review by dogezen — APPROVED** [review]

**Pull request merged**